File: python/144_Binary_Tree_Preorder_Traversal.py
#!/usr/bin/env python3

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        cur = root
        result = []

        while cur or len(stack) > 0:
            if not cur:
                cur = stack.pop(-1)
                cur = cur.right
            else:
                result.append(cur.val)
                stack.append(cur)
                cur = cur.left

        return result

File: java/src/brand_new/_144_Binary_Tree_Preorder_Traversal.java
package brand_new;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class _144_Binary_Tree_Preorder_Traversal {
	public List<Integer> preorderTraversal(TreeNode root) {
		List<Integer> result = new ArrayList<>();
		TreeNode cur = root;
		Stack<TreeNode> stack = new Stack<>();
		while (cur != null || stack.size() > 0) {
			while (cur != null) {
				result.add(cur.val);
				stack.push(cur);
				cur = cur.left;
			}

			if (stack.size() > 0) {
				cur = stack.pop();
				cur = cur.right;
			}
		}
		return result;
	}
}
